src/cargo/core/gc.rs

```diff
@@ -0,0 +1,550 @@
+            self.max_git_co_age = Some(maybe_newer_span(max_download_age, self.max_git_co_age));
+            self.max_git_db_age = Some(maybe_newer_span(max_download_age, self.max_git_db_age));
+        }
+        Ok(())
+    }
+}
+
+/// The kind of automatic garbage collection to perform.
+///
+/// "Automatic" is the kind of gc performed automatically by Cargo in any
+/// command that is already doing a bunch of work. See [`auto_gc`] for more.
+#[derive(Clone, Debug)]
+pub enum AutoGcKind {
+    /// Automatically clean up the downloaded files *and* the target directory.
+    ///
+    /// This is the mode used by default.
+    All,
+    /// Automatically clean only downloaded files.
+    ///
+    /// This corresponds to `cargo clean --gc=download`.
+    Download,
+    /// Automatically clean only the target directory.
+    ///
+    /// THIS IS NOT IMPLEMENTED.
+    ///
+    /// This corresponds to `cargo clean --gc=target`.
+    Target,
+    /// Automatically clean only the shared target directory.
+    ///
+    /// THIS IS NOT IMPLEMENTED.
+    ///
+    /// This corresponds to `cargo clean --gc=shared-target`.
+    SharedTarget,
+}
+
+impl<'a, 'config> Gc<'a, 'config> {
+    pub fn new(
+        config: &'config Config,
+        global_cache_tracker: &'a mut GlobalCacheTracker,
+    ) -> CargoResult<Gc<'a, 'config>> {
+        let lock = config.acquire_package_cache_lock(CacheLockMode::MutateExclusive)?;
+        Ok(Gc {
+            config,
+            global_cache_tracker,
+            lock,
+        })
+    }
+
+    /// Performs automatic garbage cleaning.
+    ///
+    /// This returns immediately without doing work if garbage collection has
+    /// been performed recently (since `gc.auto.frequency`).
+    fn auto(&mut self, clean_ctx: &mut CleanContext<'config>) -> CargoResult<()> {
+        if !self.config.cli_unstable().gc {
+            return Ok(());
+        }
+        let auto_config = self
+            .config
+            .get::<Option<AutoConfig>>("gc.auto")?
+            .unwrap_or_default();
+        let Some(freq) = parse_frequency(auto_config.frequency.as_deref().unwrap_or("1 day"))?
+        else {
+            tracing::trace!("auto gc disabled");
+            return Ok(());
+        };
+        if !self.global_cache_tracker.should_run_auto_gc(freq)? {
+            return Ok(());
+        }
+        let mut gc_opts = GcOpts::default();
+        gc_opts.update_for_auto_gc_config(&auto_config, &[AutoGcKind::All], None)?;
+        self.gc(clean_ctx, &gc_opts)?;
+        if !clean_ctx.dry_run {
+            self.global_cache_tracker.set_last_auto_gc()?;
+        }
+        Ok(())
+    }
+
+    /// Performs garbage collection based on the given options.
+    pub fn gc(
+        &mut self,
+        clean_ctx: &mut CleanContext<'config>,
+        gc_opts: &GcOpts,
+    ) -> CargoResult<()> {
+        self.global_cache_tracker.clean(clean_ctx, gc_opts)?;
```

Comment:
From an API perceptive, this feels weird that `GlobalCacheTracker` sometimes represents dry-run and sometimes the caller has to take care of it

Reply:
The asymmetry between
```rust
        if !clean_ctx.dry_run {
            self.global_cache_tracker.set_last_auto_gc()?;
        }
```
and
```rust
self.global_cache_tracker.clean(clean_ctx, gc_opts)?;
```
is what I'm calling out

If the API doesn't predictably handle dry_run, that will more likely lead to bugs.

Reply:
Hm, I'm not sure I see a better way to particularly do that. The `dry_run` flag could be passed to `set_last_auto_gc` instead of checking it outside, but I don't quite see how that would help. Did you have any specific ideas on what could be different there?

Reply:
I think I'd prefer a `dry_run` flag for `set_last_auto_gc` but it is a messy situation that my preference isn't strong enough to hold this back.